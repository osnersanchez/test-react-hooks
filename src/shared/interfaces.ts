export interface Gif {
    id: string;
    title: string;
    url: string;
}

export interface ServiceRespond<T> {
    data: T[],
    loading: boolean
}
