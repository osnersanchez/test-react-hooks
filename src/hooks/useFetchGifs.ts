import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";
import { Gif, ServiceRespond } from "../shared/interfaces";

export const useFetchGif = (category: string) => {
  const [state, setState] = useState<ServiceRespond<Gif>>({
    data: [],
    loading: true
  });

  useEffect(() => {
    getGifs(category)
    .then((gifList: Gif[]) => {

      setState({
        data: gifList,
        loading: false
      })
    });
  }, [category]);

  return state;
};