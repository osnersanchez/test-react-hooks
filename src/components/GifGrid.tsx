import React from 'react';
import { useFetchGif } from '../hooks/useFetchGifs';
import { Gif } from '../shared/interfaces';
import GifGridItem from './GifGridItem';

interface PropsTypes {
  category: string;
}

const GifGrid = ({ category }: PropsTypes) => {
  const { data: gifs, loading } = useFetchGif(category);

  return (
    <>
      <h3>{category}</h3>
      {loading && <p className="animate__animated animate__flash">Loading...</p>}
      <div className="card-grid">
        {
          gifs.map((gif: Gif) => (
            <GifGridItem key={gif.id} gif={gif} />
          ))
        }
      </div>
    </>
  )
};

export default GifGrid;