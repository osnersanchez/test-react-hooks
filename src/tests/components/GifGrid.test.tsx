import React from 'react';
import '@testing-library/jest-dom';
import { shallow, ShallowWrapper } from 'enzyme';

import GifGrid from '../../components/GifGrid';
import { useFetchGif } from '../../hooks/useFetchGifs';
import { Gif, ServiceRespond } from '../../shared/interfaces';

jest.mock('../../hooks/useFetchGifs');

const mockData = [{
  id: '1',
  title: 'Mygif',
  url: 'Any Url'
},{
  id: '2',
  title: 'Mygif',
  url: 'Any Url'
}];

describe('GifGrid test suite', () => {
  const mockFetch = useFetchGif as jest.Mock<ServiceRespond<Gif>>


  test('should render default template when default properties are set', () => {
    mockFetch.mockReturnValue({
      data: [],
      loading: true
    });
    const wrapper: ShallowWrapper = shallow(<GifGrid category='The Pepe' />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render default template when default properties are set', () => {
    mockFetch.mockReturnValue({
      data: mockData,
      loading: false
    });
    const wrapper: ShallowWrapper = shallow(<GifGrid category='The Pepe' />);
    expect(wrapper.find('p').exists()).toBeFalsy();
    expect(wrapper.find('GifGridItem').length).toBe(mockData.length);
  });

  //   test('should render in the p tag the title of props', () => {
  //     const pRef = wrapper.find('p');
  //     expect(pRef.text().trim()).toBe(gif.title);
  //   })

  //   test('should render in the img tag the url of props', () => {
  //     const imgRef = wrapper.find('img');
  //     expect(imgRef.prop('src')).toBe(gif.url);
  //     expect(imgRef.prop('alt')).toBe(gif.title);
  //   })

  //   test('should include the animate__bounceIn class in the div with the card-item class', () => {
  //     const divRef = wrapper.find('.card-item');
  //     expect(divRef.prop('className')?.includes('animate__bounceIn')).toBeTruthy();
  //   })
});