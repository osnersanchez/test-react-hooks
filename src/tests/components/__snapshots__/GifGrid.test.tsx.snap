// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid test suite should render default template when default properties are set 1`] = `
<Fragment>
  <h3>
    The Pepe
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
